.github/workflows/ci.yml

```diff
@@ -11,209 +11,239 @@ concurrency:
+  #        run: python -m venv .venv
+  #      - name: Run tests
+  #        run: source .venv/bin/activate && ./ci.sh
+  #      - if: always()
+  #        uses: codecov/codecov-action@v3
+  #        with:
+  #          directory: empty
+  #          token: 87cefb17-c44b-4f2f-8b30-1fff5769ce46
+  #          name: Alpine
+  #          flags: Alpine,3.12
 
-  Ubuntu:
-    name: 'Ubuntu (${{ matrix.python }}${{ matrix.extra_name }})'
-    timeout-minutes: 10
-    runs-on: 'ubuntu-latest'
-    strategy:
-      fail-fast: false
-      matrix:
-        python: ['pypy-3.9', 'pypy-3.10', '3.8', '3.9', '3.10', '3.11', '3.12', 'pypy-3.9-nightly', 'pypy-3.10-nightly']
-        check_formatting: ['0']
-        no_test_requirements: ['0']
-        extra_name: ['']
-        include:
-          - python: '3.8'
-            check_formatting: '1'
-            extra_name: ', check formatting'
-          # separate test run that doesn't install test-requirements.txt
-          - python: '3.8'
-            no_test_requirements: '1'
-            extra_name: ', no test-requirements'
-    continue-on-error: >-
-      ${{
-        (
-          endsWith(matrix.python, '-dev')
-          || endsWith(matrix.python, '-nightly')
-        )
-        && true
-        || false
-      }}
-    steps:
-      - name: Checkout
-        uses: actions/checkout@v3
-      - name: Setup python
-        uses: actions/setup-python@v4
-        if: "!endsWith(matrix.python, '-dev')"
-        with:
-          python-version: ${{ fromJSON(format('["{0}", "{1}"]', format('{0}.0-alpha - {0}.X', matrix.python), matrix.python))[startsWith(matrix.python, 'pypy')] }}
-          cache: pip
-          cache-dependency-path: test-requirements.txt
-      - name: Setup python (dev)
-        uses: deadsnakes/action@v2.0.2
-        if: endsWith(matrix.python, '-dev')
-        with:
-          python-version: '${{ matrix.python }}'
-      - name: Run tests
-        run: ./ci.sh
-        env:
-          CHECK_FORMATTING: '${{ matrix.check_formatting }}'
-          NO_TEST_REQUIREMENTS: '${{ matrix.no_test_requirements }}'
-      - if: always()
-        uses: codecov/codecov-action@v3
-        with:
-          directory: empty
-          token: 87cefb17-c44b-4f2f-8b30-1fff5769ce46
-          name: Ubuntu (${{ matrix.python }}${{ matrix.extra_name }})
-          flags: Ubuntu,${{ matrix.python }}
-
-  macOS:
-    name: 'macOS (${{ matrix.python }})'
-    timeout-minutes: 15
-    runs-on: 'macos-latest'
+  Cython:
+    name: "Cython"
+    runs-on: ubuntu-latest
     strategy:
       fail-fast: false
       matrix:
-        python: ['3.8', '3.9', '3.10', 'pypy-3.9-nightly', 'pypy-3.10-nightly']
-    continue-on-error: >-
-      ${{
-        (
-          endsWith(matrix.python, '-dev')
-          || endsWith(matrix.python, '-nightly')
-        )
-        && true
-        || false
-      }}
+        cython: ['==0.29.23', '==0.29.26', '==0.29.30', '<3', '>=3']
+        python: ['3.8', '3.10', '3.11', '3']
```

Comment:
Bit confused about versions here:
- weren't you just talking about testing `<3` and `>=3`?
- Why are we testing `'3'`? (and how about 3.12?)
- What about 3.9? (wondering if there's a reason other than matrix size -- I don't really care if 3.9's tested or not)

Reply:
sorry, should've clarified with a comment. These are just a semi-random set of versions to test/show different behaviour in the CI, not the versions I'm actually suggesting.

Reply:
specifying `'3'` could make sense if we only bother trying on two python versions, oldest and newest. So like
```
cython: ['<3', '>=3']
python: ['3.8', '3']
```
but given that all other CI steps need to get updated on highest-version bumps that's probably silly optimization.

Reply:
Yeah I'd personally rather things be similar (e.g. either all use `'3.12'` or all use `'3'`)